File: webhooks/kafka-trigger/java-spring/src/main/java/planvault/examples/kafkawebhook/WebhookSignature.java
// SPDX-License-Identifier: Apache-2.0
package planvault.examples.kafkawebhook;

import java.nio.charset.StandardCharsets;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * HMAC-SHA256 signing for PlanVault inbound webhooks ({@code hmac_sha256} trigger auth).
 *
 * <p>PlanVault verifies a hex digest over canonical JSON bytes; workers should POST the same UTF-8
 * string they sign (compact JSON). See {@code planvault-examples/webhooks/kafka-trigger/README.md}.
 */
public final class WebhookSignature {

  private WebhookSignature() {}

  /** Lowercase hex HMAC-SHA256 of {@code body} using {@code secretUtf8} as the key. */
  public static String hmacSha256Hex(byte[] secretUtf8, String body) {
    try {
      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(new SecretKeySpec(secretUtf8, "HmacSHA256"));
      byte[] raw = mac.doFinal(body.getBytes(StandardCharsets.UTF_8));
      StringBuilder sb = new StringBuilder(raw.length * 2);
      for (byte b : raw) {
        sb.append(String.format("%02x", b));
      }
      return sb.toString();
    } catch (Exception e) {
      throw new IllegalStateException(e);
    }
  }
}

File: webhooks/kafka-trigger/scala-fs2/src/main/scala/planvault/examples/kafkawebhook/WebhookSignature.scala
// SPDX-License-Identifier: Apache-2.0
package planvault.examples.kafkawebhook

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** HMAC signing for PlanVault inbound webhooks (`hmac_sha256` trigger auth).
  *
  * The server recomputes the signature over Circe's `body.noSpaces` UTF-8 bytes. For typical compact
  * JSON, signing the exact POST body bytes matches that value—see project docs and worker READMEs.
  */
object WebhookSignature {

  /** Hex-encoded HMAC-SHA256 of `body` using `secret` as UTF-8 key bytes. */
  def hmacSha256Hex(secret: String, body: String): String = {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(new SecretKeySpec(secret.getBytes("UTF-8"), "HmacSHA256"))
    mac.doFinal(body.getBytes("UTF-8")).map("%02x".format(_)).mkString
  }
}
